### ISSUE TITLE
Whether nginx shared dict 'lpop' command delete the key of list or not?

### ISSUE THREAD

**Issue by bigzuo** [openresty/lua-nginx-module#1677]
Openresty version: 1.13.6.2

According to our business needs, we need to create a list in nginx shared dict every second and put some elements into it, then there is a timer thread in nginx consumes the elements in the list with "lpop" command.
 In our test environment, after "lpop" all the elements in the list, the key of the list will be "disappeared" automatically. But in our product environment, the key is still there even though the value of the list is empty. Since we will create a list in shared dict every second, it results in OOM in the specific shared dict.
We don't find any conclusions about the behavior of key of list data structure in shared dict in openresty official website. Does anyone know why this happened?

**Comment by rainingmaster** [comment]
According the [ngx_http_lua_shdict_pop_helper](https://github.com/openresty/lua-nginx-module/tree/master/src/ngx_http_lua_shdict.c),  when the list is empty,  `lpop` should return a pair of nil and error,  is there some differences in your live environment?  Could you show the code and the result here?

**Comment by rainingmaster** [comment]
> According the [ngx_http_lua_shdict_pop_helper](https://github.com/openresty/lua-nginx-module/tree/master/src/ngx_http_lua_shdict.c),  when the list is empty,  `lpop` should return a pair of nil and error,  is there some differences in your live environment?  Could you show the code and the result here? 

Or you can use`llen` to check the length of the list?

**Comment by bigzuo** [comment]
Yeah, just as what you've said, first of all, I get the length of the list, then I pop out all the elements in the list in a `for .. do` code block. Since I didn't try to pop out elements from an empty list, so there's no error occurs with `pop` command in my code. 
But the thing is: After pop out all the elements, the key of the list will be "disappeared" automatically in our test environment. But in our product environment, the key is still there after we did the same thing. So my question is about the behavior of the key when a list is totally empty, not the `pop` command.

Here is my code:
```
local len, err = shared_dict:llen(list_key)
if len ~= 0 then
    for i = 1, len do
        local value, lpop_err = shared_dict:lpop(list_key)
        -- do something with the value
    end
end
```
After executing the code block above, the `list_key` will "disappeared" in our test environment, but not in our product environment.

**Comment by doujiang24** [comment]
@bigzuo 
the `list_key` should be disappeared when the list is empty, it's by design.
I'm wondering if there is someone other in another worker process pushing value into the `list_key` during the `lpop` in the for loop?

**Comment by rainingmaster** [comment]
@bigzuo 
Maybe you can run the below code, check is it `0` len in the list?
```
local len, err = shared_dict:llen(list_key)
if len ~= 0 then
    for i = 1, len do
        local value, lpop_err = shared_dict:lpop(list_key)
        -- do something with the value
    end
    ngx.log(ngx.ERR, shared_dict:llen(list_key))
end
```

By the way, shdict is share with all worker, it maybe like @doujiang24 said, other worker has push value into it, can you log both in `push` and `pop` to check it?

**Comment by bigzuo** [comment]
> @bigzuo
> the `list_key` should be disappeared when the list is empty, it's by design.
> I'm wondering if there is someone other in another worker process pushing value into the `list_key` during the `lpop` in the for loop?

According to our design, I can guarantee there's no other processes pushing value into the `list_key` during the `lpop` in the for loop. But in order to make sure the list is empty and no other processed push value into it, I will add `ngx.log(ngx.ERR, shared_dict:llen(list_key))` code block to  our product environment.

I will tell you the result when it come out. Thank you very much.
